File: app/src/ai-agent/message-context.ts
import { and, asc, eq, sql } from 'drizzle-orm'
import { ms } from '@pleisto/active-support'
import type { TextContent } from '@/llm'
import { DB, type QueryExecutor } from '@/common/database'
import { AiAgentMessages, type JsonObject } from '@/common/db-schema'
import { stringFromPath, toJsonObject } from '@/common/json'
import { zonedParts } from '@/config/system'
import type { AgentMessage } from './core'

// Ambient scene facts (who is speaking, in which room, at what time, the trigger reasoning) are prepended
// to a user message as a small `<message_context>` block so the model knows the situation without it
// being baked into the message body. The guiding principle is SPARSE injection: each fact is emitted only
// when it CHANGED since the last message — repeating an unchanged room/speaker/timestamp on every turn
// would burn tokens and train the model to ignore the block. The decision (`injected: true/false`) is
// computed once at write time and frozen in the message metadata, so a later re-render reproduces the
// exact same prefix instead of re-deciding against a now-different history.

export const MESSAGE_CONTEXT_METADATA_KEY = 'message_context'

// A timestamp is re-injected only after at least this much wall-clock gap, so the model is reminded of
// "now" when a conversation resumes after a pause, but not on every rapid back-and-forth message.
const TIME_CONTEXT_GAP_MS = ms('1h')
// Hard cap on any single rendered context line, defending the prompt against an oversized display
// name / room title / think string.
const MAX_CONTEXT_LINE_TEXT = 800

export interface MessageContextInput {
  actor?: JsonObject
  room?: JsonObject
  sentAt: Date
  speaker?: string
  speakerRole?: string
  speakerTrigger?: string
  think?: string
  timezone: string
}

export interface MessageContextHistoryItem {
  metadata: JsonObject
}

/**
 * Loads the prior incoming-message metadata that the change-detection below compares against.
 *
 * Restricted to the messages that actually carry a `message_context` decision: real inbound user / ambient
 * messages of a normal-or-introspection kind. Rows with a `transcript_effect` (edits, redactions, and
 * other rewrites of history) are excluded so their synthetic context does not skew "what did the last real
 * message show". Ordered oldest-first so the helpers can scan from the tail for the most recent value.
 */
export async function loadMessageContextHistory(
  conversationId: string,
  db: QueryExecutor = DB
): Promise<MessageContextHistoryItem[]> {
  const rows = await db
    .select({ metadata: AiAgentMessages.metadata })
    .from(AiAgentMessages)
    .where(
      and(
        eq(AiAgentMessages.conversationId, conversationId),
        sql`${AiAgentMessages.role} in ('user', 'im_ambient')`,
        sql`${AiAgentMessages.kind} in ('normal', 'introspection')`,
        sql`${AiAgentMessages.metadata}->'transcript_effect' is null`
      )
    )
    .orderBy(asc(AiAgentMessages.createdAt), asc(AiAgentMessages.id))

  return rows.map(row => ({ metadata: row.metadata }))
}

/**
 * Computes the frozen `message_context` metadata for one incoming message: the full set of scene facts,
 * each tagged with whether it should actually be rendered (`injected`).
 *
 * Both the value and the inject/skip decision are stored. Persisting the decision — rather than just the
 * raw facts — is what lets {@link renderMessageContextPrefix} reproduce the same prefix on every later
 * render without re-running change detection against a history that has since moved on.
 *
 * `speaker`/`think` are introspection-trigger fields supplied by the system itself; they are always
 * trusted and always injected. `time`/`room`/`actor` are sparse: each is injected only when it changed
 * versus `history` (and actor only in group rooms, where "who is talking" is ambiguous).
 */
export function buildMessageContextMetadata(
  input: MessageContextInput,
  history: MessageContextHistoryItem[]
): JsonObject {
  const sentAt = input.sentAt
  const actor = actorContext(input.actor)
  const room = roomContext(input.room, actor.displayName)
  const context: JsonObject = {
    time: {
      sent_at: sentAt.toISOString(),
      injected: shouldInjectTime(sentAt, history),
      gap_ms: TIME_CONTEXT_GAP_MS,
      timezone: input.timezone
    }
  }

  if (room) {
    context.room = {
      id: room.id ?? null,
      is_dm: room.isDM,
      name: room.name ?? null,
      label: room.label,
      injected: shouldInjectRoom(room, history)
    }
  }

  if (actor.displayName || actor.key) {
    context.actor = {
      actor_key: actor.key ?? null,
      display_name: actor.displayName ?? null,
      // Who-is-speaking only matters in a group room; in a DM the sender is unambiguous, so it is never
      // injected there. In a group it is injected only when the speaker changed from the prior message.
      injected: Boolean(room && !room.isDM && shouldInjectActor(actor, history))
    }
  }

  const speaker = normalizeContextLine(input.speaker)
  const speakerRole = normalizeContextLine(input.speakerRole)
  const speakerTrigger = normalizeContextLine(input.speakerTrigger)
  if (speaker || speakerRole || speakerTrigger) {
    context.speaker = {
      display_name: speaker ?? null,
      role: speakerRole ?? null,
      trigger: speakerTrigger ?? null,
      injected: true
    }
  }

  const think = normalizeContextLine(input.think)
  if (think) {
    context.think = {
      text: think,
      injected: true
    }
  }

  return context
}

/** Stores the computed context under its reserved key without disturbing the rest of a message's metadata. */
export function mergeMessageContextMetadata(metadata: JsonObject | undefined, context: JsonObject): JsonObject {
  return {
    ...metadata,
    [MESSAGE_CONTEXT_METADATA_KEY]: context
  }
}

/**
 * Appends a just-built context to the in-memory history so the NEXT message in the same batch compares
 * against it. Used when several incoming messages are processed in one pass, before any of them is
 * persisted and reloadable via {@link loadMessageContextHistory}.
 */
export function appendMessageContextHistory(history: MessageContextHistoryItem[], metadata: JsonObject): void {
  history.push({ metadata })
}

/**
 * Prepends the rendered `<message_context>` block to a user message for the model-bound view.
 *
 * Only user messages get a prefix (the block describes an inbound message), and only when something is
 * actually injectable. The prefix is woven into the FIRST text block — not pushed as a separate leading
 * block — so providers that expect a user turn to start with text are not surprised by, e.g., a leading
 * image. Returns the message untouched when it is not a user turn or there is nothing to inject. Operates
 * on a copy; the stored message is never mutated.
 */
export function renderMessageWithContext(message: AgentMessage, metadata: JsonObject): AgentMessage {
  if (message.role !== 'user') return message
  const prefix = renderMessageContextPrefix(metadata)
  if (!prefix) return message

  if (typeof message.content === 'string') {
    return {
      ...message,
      content: `${prefix}\n\n${message.content}`
    }
  }

  const content = [...message.content]
  const firstTextIndex = content.findIndex(block => block.type === 'text')
  // No text block to merge into (e.g. an image-only message): insert the prefix as a new leading text
  // block instead.
  if (firstTextIndex < 0) {
    return {
      ...message,
      content: [{ type: 'text', text: prefix }, ...content]
    }
  }

  const firstText = content[firstTextIndex] as TextContent
  content[firstTextIndex] = {
    ...firstText,
    text: `${prefix}\n\n${firstText.text}`
  }

  return {
    ...message,
    content
  }
}

/**
 * Renders the frozen context metadata into the `<message_context>` text block, emitting only the lines
 * whose stored `injected` flag is set. This is a pure projection of the decisions made at write time — it
 * does no change detection of its own — so it stays deterministic across re-renders. Returns `undefined`
 * when nothing is injectable, which signals callers to add no prefix at all.
 */
export function renderMessageContextPrefix(metadata: JsonObject): string | undefined {
  const context = toJsonObject(metadata[MESSAGE_CONTEXT_METADATA_KEY])
  const lines: string[] = []

  const time = toJsonObject(context.time)
  if (time.injected === true && typeof time.sent_at === 'string') {
    const timezone = typeof time.timezone === 'string' ? time.timezone : undefined
    lines.push(`sent_at: ${formatTimestamp(time.sent_at, timezone)}`)
  }

  const actor = toJsonObject(context.actor)
  const room = toJsonObject(context.room)
  if (room.injected === true && typeof room.label === 'string') lines.push(`room: ${room.label}`)

  // The explicit introspection `speaker` (system-supplied, with role/trigger) wins over the inferred group
  // `actor`: when present it fully describes who is talking and why, so the plain actor line is redundant
  // and is only used as the fallback when there is no speaker block.
  const speaker = toJsonObject(context.speaker)
  if (speaker.injected === true) {
    if (typeof speaker.display_name === 'string') lines.push(`speaker: ${speaker.display_name}`)
    if (typeof speaker.role === 'string') lines.push(`speaker_role: ${speaker.role}`)
    if (typeof speaker.trigger === 'string') lines.push(`speaker_trigger: ${speaker.trigger}`)
  } else if (actor.injected === true && typeof actor.display_name === 'string') {
    lines.push(`speaker: ${actor.display_name}`)
  }

  const think = toJsonObject(context.think)
  if (think.injected === true && typeof think.text === 'string') lines.push(`think: ${think.text}`)

  return lines.length > 0 ? `<message_context>\n${lines.join('\n')}\n</message_context>` : undefined
}

// Injects a timestamp only when the gap since the last message that carried one is at least
// TIME_CONTEXT_GAP_MS. The first-ever message has no prior time, so it is intentionally NOT injected —
// there is no elapsed gap to convey yet.
function shouldInjectTime(sentAt: Date, history: MessageContextHistoryItem[]): boolean {
  const previous = findLastContext(history, context => {
    const time = toJsonObject(context.time)
    if (typeof time.sent_at !== 'string') return undefined
    const parsed = new Date(time.sent_at)
    return Number.isNaN(parsed.getTime()) ? undefined : parsed
  })
  return Boolean(previous && sentAt.getTime() - previous.getTime() >= TIME_CONTEXT_GAP_MS)
}

// Injects the room only when it differs from the last room actually shown. Compared against the most
// recent INJECTED room (not merely the last message's room) so the model's view of "current room" is the
// reference; a never-before-shown room (no previous) is always injected.
function shouldInjectRoom(room: RoomContext, history: MessageContextHistoryItem[]): boolean {
  const previous = findLastContext(history, context => {
    const value = toJsonObject(context.room)
    if (value.injected !== true) return undefined
    const id = typeof value.id === 'string' ? value.id : undefined
    const label = typeof value.label === 'string' ? value.label : undefined
    return id || label ? { id, label } : undefined
  })
  return !previous || previous.id !== room.id || previous.label !== room.label
}

// Injects the speaker only when it changed from the IMMEDIATELY preceding message (not the last injected
// one). Identity is keyed on the stable actor key when available, falling back to display name. A first
// message (no prior context) always injects.
function shouldInjectActor(actor: ActorContext, history: MessageContextHistoryItem[]): boolean {
  const current = actor.key ?? actor.displayName
  if (!current) return false

  const previousContext = lastMessageContext(history)
  if (!previousContext) return true
  const value = toJsonObject(previousContext.actor)
  const previous = (typeof value.actor_key === 'string' && value.actor_key) || stringFromNullable(value.display_name)
  return previous !== current
}

// The single most recent message's context. Note this returns on the FIRST iteration: it deliberately
// reads only the last entry (used by actor detection, which compares against the immediately prior
// message rather than scanning back for the last injected value).
function lastMessageContext(history: MessageContextHistoryItem[]): JsonObject | undefined {
  for (let index = history.length - 1; index >= 0; index -= 1) {
    return toJsonObject(history[index]!.metadata[MESSAGE_CONTEXT_METADATA_KEY])
  }
  return undefined
}

// Scans history newest-first and returns the first value `pick` extracts, skipping messages for which it
// returns undefined. This is how time/room detection finds the last message that actually carried (or
// showed) the fact, rather than just the previous message.
function findLastContext<T>(
  history: MessageContextHistoryItem[],
  pick: (context: JsonObject) => T | undefined
): T | undefined {
  for (let index = history.length - 1; index >= 0; index -= 1) {
    const context = toJsonObject(history[index]!.metadata[MESSAGE_CONTEXT_METADATA_KEY])
    const value = pick(context)
    if (value !== undefined) return value
  }
  return undefined
}

interface ActorContext {
  displayName?: string
  key?: string
}

/**
 * Normalizes a raw actor object from any channel adapter into a stable `{key, displayName}`.
 *
 * Adapters disagree on field names, so id and name are each resolved by trying a priority list of known
 * aliases (camelCase and snake_case). The display name falls back to the key when no human-readable name
 * exists, and both are length-capped so a hostile or huge value cannot bloat the context line.
 */
function actorContext(actor: JsonObject | undefined): ActorContext {
  const source = toJsonObject(actor ?? {})
  const key =
    stringFromPath(source, ['userId']) ??
    stringFromPath(source, ['external_account_id']) ??
    stringFromPath(source, ['id']) ??
    stringFromPath(source, ['open_id'])
  const displayName =
    stringFromPath(source, ['fullName']) ??
    stringFromPath(source, ['userName']) ??
    stringFromPath(source, ['display_name']) ??
    stringFromPath(source, ['name']) ??
    key

  return {
    key: key ? key.slice(0, 160) : undefined,
    displayName: displayName ? displayName.slice(0, 160) : undefined
  }
}

interface RoomContext {
  id?: string
  isDM: boolean
  label: string
  name?: string
}

/**
 * Normalizes a raw room object into a human-readable `RoomContext`, or `undefined` when there is nothing
 * identifiable to describe. Builds a natural-language label the model reads directly ("direct message
 * with Alice", `group chat "Ops"`), preferring a DM-with-person phrasing, then a named group, then a
 * bare-id group as the last fallback.
 */
function roomContext(room: JsonObject | undefined, actorDisplayName: string | undefined): RoomContext | undefined {
  const source = toJsonObject(room ?? {})
  const id = stringFromPath(source, ['id'])
  const name = stringFromPath(source, ['name'])
  const isDM = source.isDM === true || source.is_dm === true
  if (!id && !name && !actorDisplayName) return undefined

  const label = isDM
    ? `direct message with ${actorDisplayName ?? name ?? id ?? 'unknown user'}`
    : name
      ? `group chat "${name}"`
      : `group chat ${id ?? 'unknown'}`

  return {
    id,
    isDM,
    label,
    name
  }
}

// Collapses internal whitespace to single spaces and caps the length, so multi-line or padded input
// renders as one tidy context line. Returns undefined for empty/whitespace-only input so callers can omit
// the field entirely.
function normalizeContextLine(value: string | undefined): string | undefined {
  const text = value?.trim().replace(/\s+/g, ' ').slice(0, MAX_CONTEXT_LINE_TEXT)
  return text || undefined
}

// Renders the stored UTC instant in the conversation's own timezone so the model reads a local wall-clock
// time. Falls back to the raw ISO string when the timestamp is unparseable or no timezone is known,
// rather than guessing.
function formatTimestamp(value: string, timezone?: string): string {
  const parsed = new Date(value)
  if (Number.isNaN(parsed.getTime()) || !timezone) return value
  const parts = zonedParts(timezone, parsed)
  return `${pad(parts.year, 4)}-${pad(parts.month)}-${pad(parts.day)} ${pad(parts.hour)}:${pad(parts.minute)}:${pad(parts.second)} (${timezone})`
}

function pad(value: number, length = 2): string {
  return value.toString().padStart(length, '0')
}

function stringFromNullable(value: unknown): string | undefined {
  return typeof value === 'string' && value.length > 0 ? value : undefined
}

File: app/control_plane/lib/ankole/ai_agent/message_context.ex
defmodule Ankole.AIAgent.MessageContext do
  @moduledoc """
  Builds the frozen `<message_context>` metadata stored on inbound transcript rows.

  The renderer in Agent Computer must be able to replay the exact context prefix a
  message had when it was written. For that reason this module stores both the
  raw scene facts and each sparse-injection decision in message metadata instead
  of recomputing them from whatever the transcript looks like later.
  """

  import Ecto.Query, warn: false

  alias Ankole.AIAgent.Schemas.Message

  @metadata_key "message_context"
  # "Sparse injection": scene facts (time, room, actor) are rendered into the
  # prompt only when they have changed since the model last saw them, to avoid
  # repeating a timestamp/room banner on every message. Time re-injects only
  # after a 1h gap; below that the previous time line is still considered fresh.
  @time_context_gap_ms 60 * 60 * 1_000
  # Hard cap on any single injected context line (speaker, room, think). Bounds
  # how much untrusted chat metadata can bloat the frozen prompt prefix.
  @max_context_line_text 800

  @type history_item :: %{metadata: map()}
  @type input :: %{
          optional(:actor) => map() | nil,
          optional(:room) => map() | nil,
          optional(:sent_at) => DateTime.t() | String.t() | nil,
          optional(:speaker) => String.t() | nil,
          optional(:speaker_role) => String.t() | nil,
          optional(:speaker_trigger) => String.t() | nil,
          optional(:think) => String.t() | nil,
          optional(:timezone) => String.t() | nil
        }

  @doc """
  Loads prior inbound context metadata for sparse-injection comparisons.

  Only user and ambient rows that can carry real inbound scene facts are included.
  Rows marked with `transcript_effect` are omitted so recalls/deletions do not
  skew later "what did the model last see?" decisions.
  """
  @spec load_history(module(), Ecto.UUID.t()) :: [history_item()]
  def load_history(repo, conversation_id) do
    Message
    |> where([message], message.conversation_id == ^conversation_id)
    |> where([message], message.role in ["user", "im_ambient"])
    |> where([message], message.kind in ["normal", "introspection"])
    |> where([message], message.status == "complete")
    |> where([message], fragment("?->'transcript_effect' is null", message.metadata))
    |> order_by([message], asc: message.inserted_at, asc: message.id)
    |> select([message], %{metadata: message.metadata})
    |> repo.all()
  end

  @doc """
  Computes frozen message-context metadata for one incoming message.

  Time, room, and inferred actor are sparse and compare against `history`.
  Runtime-supplied speaker and think fields are always injected because they are
  explicit control-plane facts, not facts inferred from the surrounding chat.
  """
  @spec build(input(), [history_item()]) :: map()
  def build(input, history) when is_map(input) and is_list(history) do
    sent_at = normalize_sent_at(value(input, :sent_at)) || DateTime.utc_now(:microsecond)
    timezone = text(value(input, :timezone)) || "UTC"
    actor = actor_context(value(input, :actor))
    room = room_context(value(input, :room), actor.display_name)

    %{
      "time" => %{
        "sent_at" => DateTime.to_iso8601(sent_at),
        "injected" => should_inject_time?(sent_at, history),
        "gap_ms" => @time_context_gap_ms,
        "timezone" => timezone
      }
    }
    |> put_optional("room", room_metadata(room, history))
    |> put_optional("actor", actor_metadata(actor, room, history))
    |> put_optional("speaker", speaker_metadata(input))
    |> put_optional("think", think_metadata(input))
  end

  @doc """
  Stores a computed context under the reserved metadata key.
  """
  @spec merge(map() | nil, map()) :: map()
  def merge(metadata, context) when is_map(context) do
    (metadata || %{})
    |> Map.put(@metadata_key, context)
  end

  @doc """
  Appends just-written metadata to an in-memory history list for batch writes.
  """
  @spec append_history([history_item()], map()) :: [history_item()]
  def append_history(history, metadata) when is_list(history) and is_map(metadata),
    do: history ++ [%{metadata: metadata}]

  defp room_metadata(nil, _history), do: nil

  defp room_metadata(room, history) do
    %{
      "id" => room.id,
      "is_dm" => room.is_dm,
      "name" => room.name,
      "label" => room.label,
      "injected" => should_inject_room?(room, history)
    }
    |> reject_nil()
  end

  defp actor_metadata(actor, room, history) do
    case actor.display_name || actor.key do
      nil ->
        nil

      _value ->
        %{
          "actor_key" => actor.key,
          "display_name" => actor.display_name,
          # Only worth telling the model "who is speaking" in a group chat, and
          # only when the speaker changed. In a DM the counterpart is implicit,
          # so the actor banner is never injected.
          "injected" => not is_nil(room) and !room.is_dm and should_inject_actor?(actor, history)
        }
        |> reject_nil()
    end
  end

  defp speaker_metadata(input) do
    speaker = normalize_context_line(value(input, :speaker))
    role = normalize_context_line(value(input, :speaker_role))
    trigger = normalize_context_line(value(input, :speaker_trigger))

    case speaker || role || trigger do
      nil ->
        nil

      _value ->
        %{
          "display_name" => speaker,
          "role" => role,
          "trigger" => trigger,
          "injected" => true
        }
        |> reject_nil()
    end
  end

  defp think_metadata(input) do
    case normalize_context_line(value(input, :think)) do
      nil -> nil
      think -> %{"text" => think, "injected" => true}
    end
  end

  defp should_inject_time?(sent_at, history) do
    case find_last_context(history, fn context ->
           context
           |> map_value("time")
           |> map_value("sent_at")
           |> normalize_sent_at()
         end) do
      %DateTime{} = previous ->
        DateTime.diff(sent_at, previous, :millisecond) >= @time_context_gap_ms

      nil ->
        false
    end
  end

  defp should_inject_room?(room, history) do
    previous =
      find_last_context(history, fn context ->
        value = context |> map_value("room") |> ensure_map()

        case value["injected"] == true do
          true -> %{id: text(value["id"]), label: text(value["label"])}
          false -> nil
        end
      end)

    is_nil(previous) || previous.id != room.id || previous.label != room.label
  end

  defp should_inject_actor?(actor, history) do
    current = actor.key || actor.display_name
    previous_context = last_message_context(history)
    previous_actor = previous_context |> map_value("actor") |> ensure_map()
    previous = text(previous_actor["actor_key"]) || text(previous_actor["display_name"])

    current && previous != current
  end

  defp find_last_context(history, fun) do
    history
    |> Enum.reverse()
    |> Enum.reduce_while(nil, fn item, _acc ->
      case item
           |> Map.get(:metadata, %{})
           |> map_value(@metadata_key)
           |> ensure_map()
           |> fun.() do
        nil -> {:cont, nil}
        value -> {:halt, value}
      end
    end)
  end

  defp last_message_context([]), do: %{}

  defp last_message_context(history) do
    history
    |> List.last()
    |> Map.get(:metadata, %{})
    |> map_value(@metadata_key)
    |> ensure_map()
  end

  defp actor_context(actor) do
    source = ensure_map(actor)

    key =
      text(source["userId"]) ||
        text(source["user_id"]) ||
        text(source["external_account_id"]) ||
        text(source["id"]) ||
        text(source["open_id"])

    display_name =
      text(source["fullName"]) ||
        text(source["full_name"]) ||
        text(source["userName"]) ||
        text(source["user_name"]) ||
        text(source["display_name"]) ||
        text(source["name"]) ||
        key

    %{
      key: slice(key, 160),
      display_name: slice(display_name, 160)
    }
  end

  defp room_context(room, actor_display_name) do
    source = ensure_map(room)
    id = text(source["id"])
    name = text(source["name"]) || text(source["title"])
    kind = text(source["kind"])
    is_dm = source["isDM"] == true || source["is_dm"] == true || kind == "im_dm"

    case id || name || actor_display_name do
      nil ->
        nil

      _value ->
        label =
          case is_dm do
            true -> "direct message with #{actor_display_name || name || id || "unknown user"}"
            false when is_binary(name) -> "group chat \"#{name}\""
            false -> "group chat #{id || "unknown"}"
          end

        %{id: id, is_dm: is_dm, label: label, name: name}
    end
  end

  defp normalize_sent_at(%DateTime{} = sent_at), do: sent_at

  defp normalize_sent_at(value) when is_binary(value) do
    case DateTime.from_iso8601(value) do
      {:ok, sent_at, _offset} -> sent_at
      _error -> nil
    end
  end

  defp normalize_sent_at(_value), do: nil

  defp normalize_context_line(value) do
    value
    |> text()
    |> case do
      nil -> nil
      value -> value |> String.replace(~r/\s+/, " ") |> slice(@max_context_line_text)
    end
  end

  defp value(map, key), do: Map.get(map, key) || Map.get(map, Atom.to_string(key))

  defp map_value(map, key) when is_map(map),
    do: Map.get(map, key) || Map.get(map, String.to_atom(key))

  defp map_value(_value, _key), do: nil

  defp ensure_map(value) when is_map(value), do: value
  defp ensure_map(_value), do: %{}

  defp text(value) when is_binary(value) do
    case String.trim(value) do
      "" -> nil
      value -> value
    end
  end

  defp text(_value), do: nil

  defp slice(nil, _length), do: nil
  defp slice(value, length), do: String.slice(value, 0, length)

  defp put_optional(map, _key, nil), do: map
  defp put_optional(map, key, value), do: Map.put(map, key, value)

  defp reject_nil(map) do
    map
    |> Enum.reject(fn {_key, value} -> is_nil(value) end)
    |> Map.new()
  end
end
